packages/react-compose/src/next/README.md

```diff
@@ -0,0 +1,226 @@
+# @fluentui/react-compose
+
+`@fluentui/react-compose` is a library of utilities for composing Fluent UI components.
+
+## A basic component walkthrough
+
+Building a recomposable component requires that we build legos; we put them together, but we let others put them together how they may need. Here's what's needed:
+
+- **State hooks** - all processing of the component should be done within one or more reusable hooks. The hooks should be allowed to manipulate the state in some reusable way, so that recomposing the component can be easily done.
+- **Render function** - a function which takes in final state and returns JSX. (e.g. `renderButton`)
+- **Component factory function** - a create function which returns `{ state, render }` parts for creating the component. This is used to scaffold a new version of the component as needed.
+- **Style hooks** - hooks which can take in styles and provide appropriate classnames to the anatomy of the component.
+
+With these building blocks, you can compose or recompose the component in numerous ways.
+
+### Simple example
+
+A render function:
+
+```jsx
+const renderButton = state => {
+  return <button {...state} />;
+};
+```
+
+A hook which can manipulate the state (defining accessibility and behaviors):
+
+```jsx
+const useButton = state => {
+  // If the button is rendered as something besides a button or anchor, make it focusable.
+  if (state.as !== 'button' && state.as !== 'a') {
```

Comment:
This should be computed based on `ref.current.tagName`

Reply:
Isn't that a circular dependency? We need to know, before render, what type we expect to render, in order to determine props. We need this for filtering (getNativeElementProps), we need it here for determining the appropriate behaviors.

Refs are resolved at mount. So assuming we did what you say, we'd need to first render, then re-render when we get our assumption wrong, so that we can render the correct attributes.

Could we simply lock typings to assume "as" is a string tag?

Reply:
root "as", at least.